File: app/src/main/java/org/ole/planet/myplanet/model/RealmAchievement.java
package org.ole.planet.myplanet.model;

import android.text.TextUtils;
import android.widget.EditText;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import org.ole.planet.myplanet.utilities.JsonUtils;
import org.ole.planet.myplanet.utilities.Utilities;

import io.realm.Realm;
import io.realm.RealmList;
import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;

public class RealmAchievement extends RealmObject {

    private RealmList<String> achievements;

    private RealmList<String> references;

    private String purpose;

    private String achievementsHeader;

    private String sendToNation;

    private String _rev;

    @PrimaryKey
    private String _id;

    private String goals;

    public static JsonObject serialize(RealmAchievement sub) {
        JsonObject object = new JsonObject();
        object.addProperty("_id", sub.get_id());
        if (!TextUtils.isEmpty(sub.get_rev()))
        object.addProperty("_rev", sub.get_rev());
        object.addProperty("goals", sub.getGoals());
        object.addProperty("purpose", sub.getPurpose());
        object.addProperty("achievementsHeader", sub.getAchievementsHeader());
        object.add("references", sub.getreferencesArray());
        object.add("achievements", sub.getAchievementsArray());
        Utilities.log(new Gson().toJson(object));
        return object;
    }



    public void setAchievements(RealmList<String> achievements) {
        this.achievements = achievements;
    }

    public RealmList<String> getReferences() {
        return references;
    }

    public void setReferences(RealmList<String> references) {
        this.references = references;
    }

    public static JsonObject createReference(String name, EditText relation, EditText phone, EditText email) {
        JsonObject ob = new JsonObject();
        ob.addProperty("name", name);
        ob.addProperty("phone", phone.getText().toString());
        ob.addProperty("relationship", relation.getText().toString());
        ob.addProperty("email", email.getText().toString());
        return ob;
    }

    public String getPurpose() {
        return purpose;
    }

    public void setPurpose(String purpose) {
        this.purpose = purpose;
    }

    public String getAchievementsHeader() {
        return achievementsHeader;
    }

    public void setAchievementsHeader(String achievementsHeader) {
        this.achievementsHeader = achievementsHeader;
    }

    public String getSendToNation() {
        return sendToNation;
    }

    public void setSendToNation(String sendToNation) {
        this.sendToNation = sendToNation;
    }

    public String get_rev() {
        return _rev;
    }

    public void set_rev(String _rev) {
        this._rev = _rev;
    }

    public String get_id() {
        return _id;
    }

    public void set_id(String _id) {
        this._id = _id;
    }

    public String getGoals() {
        return goals;
    }

    public void setGoals(String goals) {
        this.goals = goals;
    }


    public static void insert(Realm mRealm, JsonObject act) {
        RealmAchievement achievement = mRealm.where(RealmAchievement.class).equalTo("_id", JsonUtils.getString("_id", act)).findFirst();
        if (achievement == null)
            achievement = mRealm.createObject(RealmAchievement.class, JsonUtils.getString("_id", act));
        achievement.set_rev(JsonUtils.getString("_rev", act));
        achievement.setPurpose(JsonUtils.getString("purpose", act));
        achievement.setGoals(JsonUtils.getString("goals", act));
        achievement.setAchievementsHeader(JsonUtils.getString("achievementsHeader", act));
        achievement.setreferences(JsonUtils.getJsonArray("references", act));
        achievement.setAchievements(JsonUtils.getJsonArray("achievements", act));
    }

    public RealmList<String> getreferences() {
        return references;
    }


    public RealmList<String> getAchievements() {
        return achievements;
    }

    public JsonArray getAchievementsArray() {
        JsonArray array = new JsonArray();
        for (String s : achievements
        ) {
            JsonElement ob = new Gson().fromJson(s, JsonElement.class);
            array.add(ob);
        }
        return array;
    }

    public JsonArray getreferencesArray() {
        JsonArray array = new JsonArray();
        for (String s : references
        ) {
            JsonElement ob = new Gson().fromJson(s, JsonElement.class);
            array.add(ob);
        }
        return array;
    }

    public void setAchievements(JsonArray ac) {
        achievements = new RealmList<String>();
        for (JsonElement el : ac) {
            String achi = new Gson().toJson(el);
            if (!achievements.contains(achi))
                achievements.add(achi);
        }
    }

    public void setreferences(JsonArray of) {
        references = new RealmList<String>();
        if (of == null)return;
            for (JsonElement el : of) {
            Utilities.log("Set references");
            String e = new Gson().toJson(el);
            if (!references.contains(e))
                references.add(e);
        }
    }
}

File: app/src/main/java/org/ole/planet/myplanet/model/RealmAchievement.kt
package org.ole.planet.myplanet.model

import android.text.TextUtils
import android.widget.EditText
import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import io.realm.Realm
import io.realm.RealmList
import io.realm.RealmObject
import io.realm.annotations.PrimaryKey
import org.ole.planet.myplanet.utils.JsonUtils
import android.util.LruCache

open class RealmAchievement : RealmObject() {
    var achievements: RealmList<String>? = null
    var references: RealmList<String>? = null
    var links: RealmList<String>? = null
    var otherInfo: RealmList<String>? = null
    var purpose: String? = null
    var achievementsHeader: String? = null
    var sendToNation: String? = null
    var _rev: String? = null
    @PrimaryKey
    var _id: String? = null
    var goals: String? = null
    var dateSortOrder: String? = null
    var createdOn: String? = null
    var username: String? = null
    var parentCode: String? = null
    var isUpdated: Boolean = false

    val achievementsArray: JsonArray
        get() = parseStringListToJsonArray(achievements)

    @io.realm.annotations.Ignore
    private var cachedReferencesArray: JsonArray? = null

    fun getReferencesArray(): JsonArray {
        return parseStringListToJsonArray(references)
    }

    val linksArray: JsonArray
        get() = parseStringListToJsonArray(links)

    val otherInfoArray: JsonArray
        get() = parseStringListToJsonArray(otherInfo)

    fun setLinks(la: JsonArray?) {
        links = RealmList()
        if (la == null) return
        for (el in la) {
            val e = JsonUtils.gson.toJson(el)
            if (links?.contains(e) != true) links?.add(e)
        }
    }

    fun setOtherInfo(oi: JsonArray?) {
        otherInfo = RealmList()
        if (oi == null) return
        for (el in oi) {
            val e = JsonUtils.gson.toJson(el)
            if (otherInfo?.contains(e) != true) otherInfo?.add(e)
        }
    }

    fun setAchievements(ac: JsonArray) {
        achievements = RealmList()
        for (el in ac) {
            val achievement = JsonUtils.gson.toJson(el)
            if (achievements?.contains(achievement) != true) {
                achievements?.add(achievement)
            }
        }
    }

    fun setReferences(of: JsonArray?) {
        cachedReferencesArray = null
        references = RealmList()
        if (of == null) return
        for (el in of) {
            val e = JsonUtils.gson.toJson(el)
            if (references?.contains(e) != true) {
                references?.add(e)
            }
        }
    }

    companion object {
        private val parsedJsonCache = LruCache<String, JsonElement>(1000)

        private fun parseStringListToJsonArray(list: RealmList<String>?): JsonArray {
            val array = JsonArray()
            for (s in list ?: emptyList()) {
                var ob = parsedJsonCache.get(s)
                if (ob == null) {
                    ob = JsonUtils.gson.fromJson(s, JsonElement::class.java)
                    parsedJsonCache.put(s, ob)
                }
                array.add(ob?.deepCopy())
            }
            return array
        }

        @JvmStatic
        fun serialize(sub: RealmAchievement): JsonObject {
            val `object` = JsonObject()
            `object`.addProperty("_id", sub._id)
            if (!TextUtils.isEmpty(sub._rev)) `object`.addProperty("_rev", sub._rev)
            `object`.addProperty("goals", sub.goals)
            `object`.addProperty("purpose", sub.purpose)
            `object`.addProperty("achievementsHeader", sub.achievementsHeader)
            `object`.addProperty("sendToNation", sub.sendToNation?.toBoolean() ?: false)
            `object`.addProperty("dateSortOrder", sub.dateSortOrder ?: "none")
            `object`.addProperty("createdOn", sub.createdOn ?: "")
            `object`.addProperty("username", sub.username ?: "")
            `object`.addProperty("parentCode", sub.parentCode ?: "")
            `object`.add("references", sub.getReferencesArray())
            `object`.add("achievements", sub.achievementsArray)
            `object`.add("links", sub.linksArray)
            `object`.add("otherInfo", sub.otherInfoArray)
            return `object`
        }

        @JvmStatic
        fun createReference(name: String?, relation: EditText, phone: EditText, email: EditText): JsonObject {
            val ob = JsonObject()
            ob.addProperty("name", name)
            ob.addProperty("phone", phone.text.toString())
            ob.addProperty("relationship", relation.text.toString())
            ob.addProperty("email", email.text.toString())
            return ob
        }

        @JvmStatic
        fun insert(mRealm: Realm, act: JsonObject?) {
            var achievement = mRealm.where(RealmAchievement::class.java)
                .equalTo("_id", JsonUtils.getString("_id", act)).findFirst()
            if (achievement == null) {
                achievement = mRealm.createObject(RealmAchievement::class.java, JsonUtils.getString("_id", act))
            }
            achievement?._rev = JsonUtils.getString("_rev", act)
            achievement?.purpose = JsonUtils.getString("purpose", act)
            achievement?.goals = JsonUtils.getString("goals", act)
            achievement?.achievementsHeader = JsonUtils.getString("achievementsHeader", act)
            achievement?.sendToNation = act?.get("sendToNation")?.asString ?: "false"
            achievement?.dateSortOrder = JsonUtils.getString("dateSortOrder", act)
            achievement?.createdOn = JsonUtils.getString("createdOn", act)
            achievement?.username = JsonUtils.getString("username", act)
            achievement?.parentCode = JsonUtils.getString("parentCode", act)
            achievement?.isUpdated = false
            achievement?.setReferences(JsonUtils.getJsonArray("references", act))
            achievement?.setAchievements(JsonUtils.getJsonArray("achievements", act))
            achievement?.setLinks(JsonUtils.getJsonArray("links", act))
            achievement?.setOtherInfo(JsonUtils.getJsonArray("otherInfo", act))
        }
    }
}
